Hi. Can you help me with `Option<&'a [u8]>`

I need to return really long CLOB. I cannot make it work if it is optional.
```rust
#[derive(Default)]
struct Test {
    vec: Vec<u8>
}
#[register]
impl BasicUdf for Test {
    type Returns<'a> = Option<&'a [u8]>;

    fn init(_cfg: &UdfCfg<Init>, _args: &ArgList<Init>) -> Result<Self, String> {
        Ok(Self::default())
    }

    fn process<'a>(&'a mut self, _cfg: &UdfCfg<Process>, _args: &ArgList<Process>, _error: Option<NonZeroU8>) -> Result<Self::Returns<'a>, ProcessError> {
        self.vec = (0..255).collect();
        Ok(Some(&self.vec[..]))
    }
}
```
I got an error
```
error[E0308]: mismatched types
  --> src/test.rs:88:1
   |
88 | #[register]
   | ^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected enum `Option<&'a [u8]>`
              found enum `Option<&[u8]>`
   = note: this error originates in the attribute macro `register` (in Nightly builds, run with -Z macro-backtrace for more info)

```
If `&'a [u8]` is not optional, it compiles just fine.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,12 +4,10 @@
 
 ## [Unreleased] - ReleaseDate
 
-### Added
-
 ### Changed
 
-### Removed
-
+- Fixed compilation for `Option<&'a [u8]>` types. Added example `mishmash` to
+  complement this
 
 
 ## [0.5.2] - 2023-03-23

--- a/udf-examples/Cargo.toml
+++ b/udf-examples/Cargo.toml
@@ -8,12 +8,14 @@ publish = false
 crate-type = ["cdylib"]
 
 [dependencies]
-udf = { path = "../udf", features = ["mock", "logging-debug"] }
 lipsum = "0.8.2"
+generic-array = "0.14.7"
+sha2 = "0.10.6"
+udf = { path = "../udf", features = ["mock", "logging-debug"] }
 uuid = { version = "1.3.0", features = ["v1", "v3", "v4", "v5", "fast-rng"] }
 
 [dev-dependencies]
-diesel = { version = "2.0.3", features = ["mysql"] }
+diesel = { version = "2.0.3", default-features = false, features = ["mysql"] }
 lazy_static = "1.4.0"
 
 [features]

--- a/udf-examples/src/lib.rs
+++ b/udf-examples/src/lib.rs
@@ -29,5 +29,6 @@ mod lipsum;
 mod log_calls;
 mod lookup;
 mod median;
+mod mishmash;
 mod sequence;
 mod sum_int;

--- a/udf-examples/src/mishmash.rs
+++ b/udf-examples/src/mishmash.rs
@@ -1,0 +1,49 @@
+//! A dummy function that combines the binary representation of all its
+//! arguments
+//!
+//! # Usage
+//!
+//! ```sql
+//! CREATE FUNCTION mismmash RETURNS string SONAME 'libudf_examples.so';
+//! SELECT lipsum(8);
+//! ```
+
+use udf::prelude::*;
+
+#[derive(Debug, Default)]
+pub struct Mishmash(Vec<u8>);
+
+#[register]
+impl BasicUdf for Mishmash {
+    type Returns<'a> = Option<&'a [u8]>;
+
+    /// We expect LIPSUM(n) or LIPSUM(n, m)
+    fn init(_cfg: &UdfCfg<Init>, _args: &ArgList<Init>) -> Result<Self, String> {
+        Ok(Self::default())
+    }
+
+    fn process<'a>(
+        &'a mut self,
+        _cfg: &UdfCfg<Process>,
+        args: &ArgList<Process>,
+        _error: Option<NonZeroU8>,
+    ) -> Result<Self::Returns<'a>, ProcessError> {
+        for arg in args {
+            match arg.value() {
+                SqlResult::String(Some(v)) => self.0.extend_from_slice(v),
+                SqlResult::Real(Some(v)) => self.0.extend_from_slice(&v.to_ne_bytes()),
+                SqlResult::Int(Some(v)) => self.0.extend_from_slice(&v.to_ne_bytes()),
+                SqlResult::Decimal(Some(v)) => self.0.extend_from_slice(v.as_bytes()),
+                other => panic!("unexpected type {other:?}"),
+            }
+        }
+
+        let res = if self.0.is_empty() {
+            None
+        } else {
+            Some(self.0.as_ref())
+        };
+
+        Ok(res)
+    }
+}

--- a/udf-examples/tests/mishmash.rs
+++ b/udf-examples/tests/mishmash.rs
@@ -1,0 +1,46 @@
+#![cfg(feature = "backend")]
+
+mod backend;
+
+use backend::get_db_connection;
+use diesel::dsl::sql;
+use diesel::prelude::*;
+use diesel::sql_types::{Binary, Nullable};
+
+const SETUP: [&str; 1] = ["create or replace function mishmash
+        returns string
+        soname 'libudf_examples.so'"];
+
+#[test]
+fn test_empty() {
+    let conn = &mut get_db_connection(&SETUP);
+
+    let res: Option<Vec<u8>> = sql::<Nullable<Binary>>("select mishmash()")
+        .get_result(conn)
+        .expect("bad result");
+
+    assert!(res.is_none());
+}
+
+#[test]
+fn test_single() {
+    let conn = &mut get_db_connection(&SETUP);
+
+    let res: Option<Vec<u8>> = sql::<Nullable<Binary>>("select mishmash('banana')")
+        .get_result(conn)
+        .expect("bad result");
+
+    assert_eq!(res.unwrap(), b"banana");
+}
+
+#[test]
+fn test_many() {
+    let conn = &mut get_db_connection(&SETUP);
+
+    let res: Option<Vec<u8>> =
+        sql::<Nullable<Binary>>("select mishmash('banana', 'is', 'a', 'fruit')")
+            .get_result(conn)
+            .expect("bad result");
+
+    assert_eq!(res.unwrap(), b"bananaisafruit");
+}

--- a/udf-macros/src/register.rs
+++ b/udf-macros/src/register.rs
@@ -181,7 +181,14 @@ fn make_basic_fns_content(
     let init_fn = make_init_fn(dstruct_ident, wrapper_ident, &init_fn_name);
     let deinit_fn = make_deinit_fn(dstruct_ident, wrapper_ident, &deinit_fn_name);
     let process_fn = match rt.type_cls {
-        TypeClass::BytesRef | TypeClass::Bytes => make_proc_buf_fn(
+        TypeClass::Bytes => make_proc_buf_fn(
+            dstruct_ident,
+            wrapper_ident,
+            &process_fn_name,
+            rt.is_optional,
+            false,
+        ),
+        TypeClass::BytesRef => make_proc_buf_fn(
             dstruct_ident,
             wrapper_ident,
             &process_fn_name,
@@ -297,9 +304,11 @@ fn make_proc_buf_fn(
     wrapper_ident: &Ident,
     fn_name: &Ident,
     is_optional: bool,
-    can_return_ref: bool,
+    is_ref: bool,
 ) -> proc_macro2::TokenStream {
-    let wrap_fn_name = if is_optional {
+    let wrap_fn_name = if is_optional && is_ref {
+        quote!(udf::wrapper::wrap_process_buf_option_ref::<#wrapper_ident, #dstruct_ident, _>)
+    } else if is_optional {
         quote!(udf::wrapper::wrap_process_buf_option::<#wrapper_ident, #dstruct_ident, _>)
     } else {
         quote!(udf::wrapper::wrap_process_buf::<#wrapper_ident, #dstruct_ident>)
@@ -322,7 +331,6 @@ fn make_proc_buf_fn(
                 length,
                 is_null,
                 error,
-                #can_return_ref,
             )
         }
     }

--- a/udf-macros/src/types.rs
+++ b/udf-macros/src/types.rs
@@ -49,8 +49,8 @@ pub fn make_type_list() -> Vec<RetType> {
         // Only valid float types
         RetType::new(parse_quote! { f64 }, false, TypeClass::Float),
         RetType::new(parse_quote! { Option<f64> }, true, TypeClass::Float),
-        // Tons of possible byte slice references. These will get copied if they
-        // fit, otherwise the reference returned
+        // Tons of possible byte slice references. We could probably make these
+        // generic somehow in the future, but it is proving tough.
         RetType::new(parse_quote! { &'a [u8] }, false, TypeClass::BytesRef),
         RetType::new(parse_quote! { Option<&'a [u8]> }, true, TypeClass::BytesRef),
         RetType::new(parse_quote! { &str }, false, TypeClass::BytesRef),

--- a/udf/src/wrapper.rs
+++ b/udf/src/wrapper.rs
@@ -13,6 +13,7 @@ pub(crate) use helpers::*;
 pub use modded_types::UDF_ARGSx;
 pub use process::{
     wrap_process_basic, wrap_process_basic_option, wrap_process_buf, wrap_process_buf_option,
+    wrap_process_buf_option_ref,
 };
 
 #[cfg(test)]

--- a/udf/src/wrapper/helpers.rs
+++ b/udf/src/wrapper/helpers.rs
@@ -1,9 +1,5 @@
 //! Private module that handles the implementation of the wrapper module
 
-// #![allow(dead_code)]
-
-use std::any::type_name;
-use std::backtrace::Backtrace;
 use std::cmp::min;
 use std::ffi::{c_char, c_ulong};
 use std::ptr;
@@ -33,19 +29,12 @@ pub unsafe fn write_msg_to_buf<const N: usize>(msg: &[u8], buf: *mut c_char) {
 pub struct BufOptions {
     res_buf: *mut c_char,
     length: *mut c_ulong,
-    /// True if we can return a reference to our source buffer. If false, we must
-    /// truncate
-    can_return_ref: bool,
 }
 
 impl BufOptions {
     /// Create a new `BufOptions` struct
-    pub fn new(res_buf: *mut c_char, length: *mut c_ulong, can_return_ref: bool) -> Self {
-        Self {
-            res_buf,
-            length,
-            can_return_ref,
-        }
+    pub fn new(res_buf: *mut c_char, length: *mut c_ulong) -> Self {
+        Self { res_buf, length }
     }
 }
 
@@ -61,10 +50,7 @@ impl BufOptions {
 ///   an error message, return None
 ///
 /// The `U` type parameter is just used for output formatting
-pub unsafe fn buf_result_callback<U, T: AsRef<[u8]>>(
-    input: T,
-    opts: &BufOptions,
-) -> Option<*const c_char> {
+pub unsafe fn buf_result_callback<T: AsRef<[u8]>>(input: T, opts: &BufOptions) -> *const c_char {
     let slice_ref = input.as_ref();
     let slice_len = slice_ref.len();
     let slice_len_ulong: c_ulong = slice_len.try_into().unwrap_or_else(|_| {
@@ -77,25 +63,12 @@ pub unsafe fn buf_result_callback<U, T: AsRef<[u8]>>(
         // If we fit in the buffer, just copy
         ptr::copy(slice_ptr, opts.res_buf, slice_len);
         *opts.length = slice_len_ulong;
-        return Some(opts.res_buf);
-    }
-
-    if !opts.can_return_ref {
-        // We can't return a reference but also can't fit in the buffer
-        *opts.length = 0;
-        udf_log!(
-            Critical: "output overflow, returning NULL. Buffer size: {}, data length: {} at `{}::process`",
-            buf_len, slice_len, type_name::<U>()
-        );
-        udf_log!(Critical: "contact your UDF vendor as this is a serious bug");
-        udf_log!(Critical: "run with `RUST_LIB_BACKTRACE=1` for a full backtrace");
-        eprintln!("{:#?}", Backtrace::capture());
-        return None;
+        return opts.res_buf;
     }
 
     // If we don't fit in the buffer but can return a reference, do so
     *opts.length = slice_len_ulong;
-    Some(slice_ptr)
+    slice_ptr
 }
 
 #[cfg(test)]
@@ -280,11 +253,9 @@ mod buffer_tests {
         let mut res_buf = [0u8; BUF_LEN];
         let zeroes = [0u8; BUF_LEN];
         let mut len = res_buf.len() as c_ulong;
-        let buf_opts = BufOptions::new(res_buf.as_mut_ptr().cast(), &mut len, false);
+        let buf_opts = BufOptions::new(res_buf.as_mut_ptr().cast(), &mut len);
 
-        let res_ptr: *const u8 = unsafe { buf_result_callback::<u8, _>(input, &buf_opts) }
-            .unwrap()
-            .cast();
+        let res_ptr: *const u8 = unsafe { buf_result_callback(input, &buf_opts).cast() };
         let res_slice = unsafe { slice::from_raw_parts(res_ptr, len as usize) };
 
         assert_eq!(len as usize, input.len());
@@ -303,11 +274,9 @@ mod buffer_tests {
         let input = b"123456789012345";
         let mut res_buf = [0u8; BUF_LEN];
         let mut len = res_buf.len() as c_ulong;
-        let buf_opts = BufOptions::new(res_buf.as_mut_ptr().cast(), &mut len, true);
+        let buf_opts = BufOptions::new(res_buf.as_mut_ptr().cast(), &mut len);
 
-        let res_ptr: *const u8 = unsafe { buf_result_callback::<u8, _>(input, &buf_opts) }
-            .unwrap()
-            .cast();
+        let res_ptr: *const u8 = unsafe { buf_result_callback(input, &buf_opts).cast() };
         let res_slice = unsafe { slice::from_raw_parts(res_ptr, len as usize) };
 
         assert_eq!(len as usize, input.len());
@@ -315,17 +284,17 @@ mod buffer_tests {
         assert_eq!(res_ptr.cast(), input.as_ptr());
     }
 
-    #[test]
-    fn test_buf_no_fit_no_ref() {
-        // Test a buffer that does not fit but can not be used as a ref
-        // This must return an error
-        let input = b"123456789012345";
-        let mut res_buf = [0u8; BUF_LEN];
-        let mut len = res_buf.len() as c_ulong;
-        let buf_opts = BufOptions::new(res_buf.as_mut_ptr().cast(), &mut len, false);
+    // #[test]
+    // fn test_buf_no_fit_no_ref() {
+    //     // Test a buffer that does not fit but can not be used as a ref
+    //     // This must return an error
+    //     let input = b"123456789012345";
+    //     let mut res_buf = [0u8; BUF_LEN];
+    //     let mut len = res_buf.len() as c_ulong;
+    //     let buf_opts = BufOptions::new(res_buf.as_mut_ptr().cast(), &mut len, false);
 
-        let res = unsafe { buf_result_callback::<u8, _>(input, &buf_opts) };
-        assert_eq!(len, 0);
-        assert_eq!(res, None);
-    }
+    //     let res = unsafe { buf_result_callback::<u8, _>(input, &buf_opts) };
+    //     assert_eq!(len, 0);
+    //     assert_eq!(res, None);
+    // }
 }

--- a/udf/src/wrapper/process.rs
+++ b/udf/src/wrapper/process.rs
@@ -124,7 +124,6 @@ pub unsafe fn wrap_process_buf<W, U>(
     length: *mut c_ulong,
     is_null: *mut c_uchar,
     error: *mut c_uchar,
-    can_return_ref: bool,
 ) -> *const c_char
 where
     W: UdfConverter<U>,
@@ -139,15 +138,12 @@ where
     let err = *(error as *const Option<NonZeroU8>);
     let binding = b.as_mut_ref();
     let proc_res = U::process(binding, cfg, arglist, err);
-    let buf_opts = BufOptions::new(result, length, can_return_ref);
+    let buf_opts = BufOptions::new(result, length);
 
     let post_effects_val = ret_callback(proc_res, error, is_null);
 
     let ret = match post_effects_val {
-        Some(ref v) => buf_result_callback::<U, _>(v, &buf_opts).unwrap_or_else(|| {
-            *error = 1;
-            ptr::null()
-        }),
+        Some(ref v) => buf_result_callback(v, &buf_opts),
         None => ptr::null(),
     };
 
@@ -168,7 +164,6 @@ pub unsafe fn wrap_process_buf_option<W, U, B>(
     length: *mut c_ulong,
     is_null: *mut c_uchar,
     error: *mut c_uchar,
-    can_return_ref: bool,
 ) -> *const c_char
 where
     W: UdfConverter<U>,
@@ -182,23 +177,52 @@ where
     let err = *(error as *const Option<NonZeroU8>);
     let mut b = cfg.retrieve_box::<W>();
     let proc_res = U::process(b.as_mut_ref(), cfg, arglist, err);
-    let buf_opts = BufOptions::new(result, length, can_return_ref);
+    let buf_opts = BufOptions::new(result, length);
 
     let post_effects_val = ret_callback_option(proc_res, error, is_null);
 
     let ret = match post_effects_val {
-        Some(ref v) => {
-            if let Some(x) = buf_result_callback::<U, _>(v, &buf_opts) {
-                x
-            } else {
-                *error = 1;
-                ptr::null()
-            }
-        }
+        Some(ref v) => buf_result_callback(v, &buf_opts),
         None => ptr::null(),
     };
 
     std::mem::forget(post_effects_val);
+    cfg.store_box(b);
+
+    log_call!(exit: "process", U, &*initid, &*args, result, &*length, &*is_null, &*error, ret);
+    ret
+}
+
+#[inline]
+pub unsafe fn wrap_process_buf_option_ref<W, U, B>(
+    initid: *mut UDF_INIT,
+    args: *mut UDF_ARGS,
+    result: *mut c_char,
+    length: *mut c_ulong,
+    is_null: *mut c_uchar,
+    error: *mut c_uchar,
+) -> *const c_char
+where
+    W: UdfConverter<U>,
+    for<'a> U: BasicUdf<Returns<'a> = Option<&'a B>>,
+    B: AsRef<[u8]> + ?Sized,
+{
+    log_call!(enter: "process", U, &*initid, &*args, result, &*length, &*is_null, &*error);
+
+    let cfg = UdfCfg::from_raw_ptr(initid);
+    let arglist = ArgList::from_raw_ptr(args);
+    let err = *(error as *const Option<NonZeroU8>);
+    let mut b = cfg.retrieve_box::<W>();
+    let proc_res = U::process(b.as_mut_ref(), cfg, arglist, err);
+    let buf_opts = BufOptions::new(result, length);
+
+    let post_effects_val = ret_callback_option(proc_res, error, is_null);
+
+    let ret = match post_effects_val {
+        Some(ref v) => buf_result_callback(v, &buf_opts),
+        None => ptr::null(),
+    };
+
     cfg.store_box(b);
 
     log_call!(exit: "process", U, &*initid, &*args, result, &*length, &*is_null, &*error, ret);

--- a/udf/src/wrapper/tests.rs
+++ b/udf/src/wrapper/tests.rs
@@ -5,6 +5,7 @@ struct ExampleInt;
 struct ExampleIntOpt;
 struct ExampleBufRef;
 struct ExampleBufOpt;
+struct ExampleBufOptRef;
 
 impl BasicUdf for ExampleInt {
     type Returns<'a> = i64;
@@ -91,6 +92,23 @@ impl AggregateUdf for ExampleBufOpt {
     }
 }
 
+impl BasicUdf for ExampleBufOptRef {
+    type Returns<'a> = Option<&'a str>;
+
+    fn init(_cfg: &UdfCfg<crate::Init>, _args: &ArgList<crate::Init>) -> Result<Self, String> {
+        todo!()
+    }
+
+    fn process<'a>(
+        &'a mut self,
+        _cfg: &UdfCfg<crate::Process>,
+        _args: &ArgList<crate::Process>,
+        _error: Option<NonZeroU8>,
+    ) -> Result<Self::Returns<'a>, ProcessError> {
+        todo!()
+    }
+}
+
 #[test]
 #[should_panic]
 #[allow(unreachable_code)]
@@ -101,8 +119,8 @@ fn test_fn_sig() {
     unsafe {
         wrap_process_basic::<ExampleInt, _, _>(todo!(), todo!(), todo!(), todo!());
         wrap_process_basic_option::<ExampleIntOpt, _, _>(todo!(), todo!(), todo!(), todo!());
-        wrap_process_buf::<ExampleBufRef, _>(
-            todo!(),
+        wrap_process_buf::<ExampleBufRef, _>(todo!(), todo!(), todo!(), todo!(), todo!(), todo!());
+        wrap_process_buf_option::<ExampleBufOpt, _, _>(
             todo!(),
             todo!(),
             todo!(),
@@ -110,8 +128,7 @@ fn test_fn_sig() {
             todo!(),
             todo!(),
         );
-        wrap_process_buf_option::<ExampleBufOpt, _, _>(
-            todo!(),
+        wrap_process_buf_option_ref::<ExampleBufOptRef, _, _>(
             todo!(),
             todo!(),
             todo!(),

